### ISSUE TITLE
consider vctrs_rcrd default format method

### ISSUE THREAD

**Issue by maxheld83** [r-lib/vctrs#1326]
The vignette states:

> vctrs makes it easy to create new record-style classes using `new_rcrd()`, which has a wide selection of default methods.

But:

``` r
library(vctrs)

# works
foo <- new_vctr(1L, class = "test_qux")
class(foo)
#> [1] "test_qux"   "vctrs_vctr"

# weirdly, this already emits a warning
bar <- new_rcrd(fields = list(corge = 1), class = "test_zap")
class(bar)
#> [1] "test_zap"   "vctrs_rcrd" "vctrs_vctr"
# this errors out
bar
#> <test_zap[1]>
#> Error: `format.test_zap()` not implemented.
```

<sup>Created on 2021-02-24 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

<details style="margin-bottom:10px;">
<summary>
Standard output and standard error
</summary>

``` sh
Error: `format.test_zap()` not implemented.
Backtrace:
     █
  1. ├─base::tryCatch(...)
  2. │ └─base:::tryCatchList(expr, classes, parentenv, handlers)
  3. │   ├─base:::tryCatchOne(...)
  4. │   │ └─base:::doTryCatch(return(expr), name, parentenv, handler)
  5. │   └─base:::tryCatchList(expr, names[-nh], parentenv, handlers[-nh])
  6. │     └─base:::tryCatchOne(expr, names, parentenv, handlers[[1L]])
  7. │       └─base:::doTryCatch(return(expr), name, parentenv, handler)
  8. ├─base::withCallingHandlers(...)
  9. ├─base::saveRDS(...)
 10. ├─base::do.call(...)
 11. ├─(function (what, args, quote = FALSE, envir = parent.frame()) ...
 12. └─(function (input, opts) ...
 13.   └─rmarkdown::render(input, quiet = TRUE, envir = globalenv(), encoding = "UTF-8")
 14.     └─knitr::knit(knit_input, knit_output, envir = envir, quiet = quiet)
 15.       └─knitr:::process_file(text, output)
 16.         ├─base::withCallingHandlers(...)
 17.         ├─knitr:::process_group(group)
 18.         └─knitr:::process_group.block(group)
 19.           └─knitr:::call_block(x)
 20.             └─knitr:::block_exec(params)
 21.               ├─knitr:::in_dir(...)
 22.               └─knitr:::evaluate(...)
 23.                 └─evaluate::evaluate(...)
 24.                   └─evaluate:::evaluate_call(...)
 25.                     ├─evaluate:::handle(...)
 26.                     │ └─base::try(f, silent = TRUE)
 27.                     │   └─base::tryCatch(...)
 28.                     │     └─base:::tryCatchList(expr, classes, parentenv, handlers)
 29.                     │       └─base:::tryCatchOne(expr, names, parentenv, handlers[[1L]])
 30.                     │         └─base:::doTryCatch(return(expr), name, parentenv, handler)
 31.                     ├─base::withCallingHandlers(...)
 32.                     ├─base::withVisible(value_fun(ev$value, ev$visible))
 33.                     └─knitr:::value_fun(ev$value, ev$visible)
 34.                       └─knitr:::fun(x, options = options)
 35.                         ├─base::withVisible(knit_print(x, ...))
 36.                         ├─knitr::knit_print(x, ...)
 37.                         └─knitr:::knit_print.default(x, ...)
 38.                           └─evaluate:::normal_print(x)
 39.                             ├─base::print(x)
 40.                             └─vctrs:::print.vctrs_vctr(x)
 41.                               └─vctrs::obj_print(x, ...)
 42.                                 ├─vctrs::obj_print_data(x, ...)
 43.                                 └─vctrs:::obj_print_data.default(x, ...)
 44.                                   ├─stats::setNames(format(x), names(x))
 45.                                   ├─base::format(x)
 46.                                   └─vctrs:::format.vctrs_rcrd(x)
 47.                                     └─vctrs:::stop_unimplemented(x, "format")
 48.                                       └─vctrs:::stop_vctrs(...)
```

</details>
<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 4.0.4 (2021-02-15)
#>  os       macOS Big Sur 10.16         
#>  system   x86_64, darwin17.0          
#>  ui       X11                         
#>  language en_US.UTF-8 git             
#>  collate  en_US.UTF-8                 
#>  ctype    en_US.UTF-8                 
#>  tz       Europe/Berlin               
#>  date     2021-02-24                  
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version date       lib source        
#>  assertthat    0.2.1   2019-03-21 [1] CRAN (R 4.0.2)
#>  backports     1.2.1   2020-12-09 [1] CRAN (R 4.0.2)
#>  cli           2.3.0   2021-01-31 [1] CRAN (R 4.0.2)
#>  crayon        1.4.1   2021-02-08 [1] CRAN (R 4.0.2)
#>  digest        0.6.27  2020-10-24 [1] CRAN (R 4.0.2)
#>  ellipsis      0.3.1   2020-05-15 [1] CRAN (R 4.0.2)
#>  evaluate      0.14    2019-05-28 [1] CRAN (R 4.0.1)
#>  fs            1.5.0   2020-07-31 [1] CRAN (R 4.0.2)
#>  glue          1.4.2   2020-08-27 [1] CRAN (R 4.0.2)
#>  highr         0.8     2019-03-20 [1] CRAN (R 4.0.2)
#>  htmltools     0.5.1.1 2021-01-22 [1] CRAN (R 4.0.2)
#>  knitr         1.31    2021-01-27 [1] CRAN (R 4.0.2)
#>  lifecycle     1.0.0   2021-02-15 [1] CRAN (R 4.0.2)
#>  magrittr      2.0.1   2020-11-17 [1] CRAN (R 4.0.2)
#>  pillar        1.4.7   2020-11-20 [1] CRAN (R 4.0.2)
#>  pkgconfig     2.0.3   2019-09-22 [1] CRAN (R 4.0.2)
#>  purrr         0.3.4   2020-04-17 [1] CRAN (R 4.0.2)
#>  reprex        1.0.0   2021-01-27 [1] CRAN (R 4.0.2)
#>  rlang         0.4.10  2020-12-30 [1] CRAN (R 4.0.2)
#>  rmarkdown     2.6     2020-12-14 [1] CRAN (R 4.0.2)
#>  sessioninfo   1.1.1   2018-11-05 [1] CRAN (R 4.0.2)
#>  stringi       1.5.3   2020-09-09 [1] CRAN (R 4.0.2)
#>  stringr       1.4.0   2019-02-10 [1] CRAN (R 4.0.2)
#>  styler        1.3.2   2020-02-23 [1] CRAN (R 4.0.2)
#>  tibble        3.0.6   2021-01-29 [1] CRAN (R 4.0.2)
#>  vctrs       * 0.3.6   2020-12-17 [1] CRAN (R 4.0.2)
#>  withr         2.4.1   2021-01-26 [1] CRAN (R 4.0.2)
#>  xfun          0.21    2021-02-10 [1] CRAN (R 4.0.2)
#>  yaml          2.2.1   2020-02-01 [1] CRAN (R 4.0.2)
#> 
#> [1] /Library/Frameworks/R.framework/Versions/4.0/Resources/library
```

</details>

This seems to affect both CRAN and dev.

Didn't this use to work with a default method?

This feels like I am missing something essential, or just plain loosing this. 
I apologise in advance if this is noise, I just can't figure out what changed.

**Comment by DavisVaughan** [comment]
> # weirdly, this already emits a warning

I don't see any warning on creation

``` r
library(vctrs)

bar <- new_rcrd(fields = list(corge = 1), class = "test_zap")

bar
#> <test_zap[1]>
#> Error: `format.test_zap()` not implemented.

format.test_zap <- function(x, ...) {
  format(field(x, "corge"))
}

bar
#> <test_zap[1]>
#> [1] 1
```

<sup>Created on 2021-02-24 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

The format method varies greatly across record types, so you'll have to implement that one yourself.

**Comment by DavisVaughan** [comment]
You get things like `[`, `[[`, and a `vec_proxy()` method for free, among other things. These core bits power higher level functions, like a meaningful `vec_sort()`

``` r
library(vctrs)

bar <- new_rcrd(fields = list(corge = c(5, 5, 1), pitbull = c(2, 1, 3)), class = "test_zap")

bar
#> <test_zap[3]>
#> Error: `format.test_zap()` not implemented.

format.test_zap <- function(x, ...) {
  paste0("[c:", field(x, "corge"), " / p:", field(x, "pitbull"), "]")
}

bar
#> <test_zap[3]>
#> [1] [c:5 / p:2] [c:5 / p:1] [c:1 / p:3]

# lots for free
bar[c(1, 1, 2)]
#> <test_zap[3]>
#> [1] [c:5 / p:2] [c:5 / p:2] [c:5 / p:1]
bar[[1]]
#> <test_zap[1]>
#> [1] [c:5 / p:2]

vec_proxy(bar)
#>   corge pitbull
#> 1     5       2
#> 2     5       1
#> 3     1       3

vec_sort(bar)
#> <test_zap[3]>
#> [1] [c:1 / p:3] [c:5 / p:1] [c:5 / p:2]
```

<sup>Created on 2021-02-24 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Comment by maxheld83** [comment]
my bad, don't know why I thought there ever was a default format method.

TL;DR: It's unusual to be able (even in dev) to *construct* objects, which you can't even `str()` or `print()`. Would that warrant a default `format.vctrs_rcrd()`?

---

I'm wondering whether others might trip on this as well.

For example, if you follow along the chunks in the vignette, you *couldn't* `str()` or `print()` the new class until you implemented the format method:

``` r
library(vctrs)
library(rlang)
new_rational <- function(n = integer(), d = integer()) {
  vec_assert(n, ptype = integer())
  vec_assert(d, ptype = integer())

  new_rcrd(list(n = n, d = d), class = "vctrs_rational")
}

# during dev or learning, I might try this out
qux <- new_rational(1L, 2L)
print(qux)
#> <vctrs_rational[1]>
#> Error: `format.vctrs_rational()` not implemented.
str(qux)
#> Error: `format.vctrs_rational()` not implemented.
```

The vignette is of course correct, but I might want to look at the object during the dev *before* I have the format method.

Would you perhaps consider adding a default, which just exposes the list structure underneath, so that people remember to make a *real* `format.()`?

``` r
vctrs::new_vctr(list(foo = c(1,2), zap = "zong"), class = "test_graupl")
#> <test_graupl[2]>
#>  foo  zap 
#> 1, 2 zong
```

Or maybe alert users to this in the vignette? (Happy to do a PR).

(absolutely ❤️ your vctrs work)

**Comment by DavisVaughan** [comment]
In https://vctrs.r-lib.org/articles/s3-vector.html#rational-class-2, before "This allows us to create a format method:", it might be nice to say something like "If you try and print out a rational object at this point, you'll get an error because there isn't a `format()` method implemented for this class" with an example of attempting to print it out

A PR for that would be welcome!

**Comment by lionel-** [comment]
Working on unit tests for vctrs rcrds, I regularly found it limiting that there is no default format method.

I think it'd be nice to have a default method that shows the underlying data. It could display an advice about implementing the format method, and link to a relevant place.
